applications/IgaApplication/custom_conditions/sbm_neumann_laplacian_condition.cpp

```diff
@@ -0,0 +1,407 @@
+}
+
+void SbmNeumannLaplacianCondition::CalculateLeftHandSide(
+    MatrixType& rLeftHandSideMatrix,
+    const ProcessInfo& rCurrentProcessInfo)
+{
+    ConvectionDiffusionSettings::Pointer p_settings = rCurrentProcessInfo[CONVECTION_DIFFUSION_SETTINGS];
+    const auto& r_geometry = this->GetGeometry();
+    const SizeType number_of_nodes = r_geometry.PointsNumber();
+
+    if (rLeftHandSideMatrix.size1() != number_of_nodes || rLeftHandSideMatrix.size2() != number_of_nodes) {
+        rLeftHandSideMatrix.resize(number_of_nodes, number_of_nodes, false);
+    }
+    noalias(rLeftHandSideMatrix) = ZeroMatrix(number_of_nodes, number_of_nodes);
+
+    // Integration
+    const GeometryType::IntegrationPointsArrayType& r_integration_points = r_geometry.IntegrationPoints();
+    const GeometryType::ShapeFunctionsGradientsType& r_DN_De = r_geometry.ShapeFunctionsLocalGradients(r_geometry.GetDefaultIntegrationMethod());
+    
+    // Initialize DN_DX
+    Matrix DN_DX(number_of_nodes,mDim);
+
+    // Calculating the PHYSICAL SPACE derivatives (it is avoided storing them to minimize storage)
+    noalias(DN_DX) = r_DN_De[0]; // prod(r_DN_De[0],InvJ0);
+    
+    const Matrix& H = r_geometry.ShapeFunctionsValues();
+
+    // compute gradient Taylor expansion contribution: grad_H_sum
+    Matrix grad_H_sum = ZeroMatrix(3, number_of_nodes);
+    ComputeGradientTaylorExpansionContribution(grad_H_sum);
+    
+    // dot product grad cdot n
+    Matrix H_grad_N_dot_n = ZeroMatrix(1, number_of_nodes);
+    noalias(row(H_grad_N_dot_n, 0)) = row(grad_H_sum, 0) * mTrueNormal[0] + row(grad_H_sum, 1) * mTrueNormal[1] + row(grad_H_sum, 2) * mTrueNormal[2];
+
+    Matrix DN_dot_n_tilde = ZeroMatrix(1, number_of_nodes);
+    for (IndexType i = 0; i < number_of_nodes; ++i)
+    {
+        // grad N cdot n_tilde
+        for (IndexType idim = 0; idim < mDim; idim++) {
+            DN_dot_n_tilde(0, i)  += DN_DX(i, idim) * mNormalParameterSpace[idim];
+        } 
+    }
+    
+    // Assembly
+    // compute Neumann contributions
+    noalias(rLeftHandSideMatrix) += prod(trans(H), H_grad_N_dot_n) * mTrueDotSurrogateNormal * r_integration_points[0].Weight(); // * std::abs(determinant_jacobian_vector[point_number]) ;
+    noalias(rLeftHandSideMatrix) -= prod(trans(H), DN_dot_n_tilde)                           * r_integration_points[0].Weight() ; // * std::abs(DetJ0) ;
+}
+
+
+void SbmNeumannLaplacianCondition::CalculateRightHandSide(
+    VectorType& rRightHandSideVector,
+    const ProcessInfo& rCurrentProcessInfo)
+{
+    const SizeType mat_size = GetGeometry().size() * 1;
+
+    if (rRightHandSideVector.size() != mat_size)
+        rRightHandSideVector.resize(mat_size);
+    noalias(rRightHandSideVector) = ZeroVector(mat_size);
+
+    MatrixType left_hand_side_matrix = ZeroMatrix(mat_size, mat_size);
+
+    CalculateLocalSystem(left_hand_side_matrix, rRightHandSideVector,
+        rCurrentProcessInfo);

```

Comment:
Considering that this is probably the final implementation, I'd avoid this trick in here.